lib/PuppeteerSharp/WaitTask.cs

```diff
@@ -1,17 +1,192 @@
 ﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace PuppeteerSharp
 {
     internal class WaitTask
     {
-        internal void Rerun()
+        private readonly Frame _frame;
+        private readonly string _predicateBody;
+        private readonly string _polling;
+        private readonly int _timeout;
+        private readonly object[] _args;
+        private readonly Task _timeoutTimer;
+
+        private readonly CancellationTokenSource _cts;
+        private readonly TaskCompletionSource<JSHandle> _taskCompletion;
+
+        private int _runCount = 0;
+        private bool _terminated;
+
+        private const string WaitForPredicatePageFunction = @"
+async function waitForPredicatePageFunction(predicateBody, polling, timeout, ...args) {
+  const predicate = new Function('...args', predicateBody);
+  let timedOut = false;
+  setTimeout(() => timedOut = true, timeout);
+  if (polling === 'raf')
+    return await pollRaf();
+  if (polling === 'mutation')
+    return await pollMutation();
+  if (typeof polling === 'number')
+    return await pollInterval(polling);
+
+  /**
+   * @return {!Promise<*>}
+   */
+  function pollMutation() {
+    const success = predicate.apply(null, args);
+    if (success)
+      return Promise.resolve(success);
+
+    let fulfill;
+    const result = new Promise(x => fulfill = x);
+    const observer = new MutationObserver(mutations => {
+      if (timedOut) {
+        observer.disconnect();
+        fulfill();
+      }
+      const success = predicate.apply(null, args);
+      if (success) {
+        observer.disconnect();
+        fulfill(success);
+      }
+    });
+    observer.observe(document, {
+      childList: true,
+      subtree: true,
+      attributes: true
+    });
+    return result;
+  }
+
+  /**
+   * @return {!Promise<*>}
+   */
+  function pollRaf() {
+    let fulfill;
+    const result = new Promise(x => fulfill = x);
+    onRaf();
+    return result;
+
+    function onRaf() {
+      if (timedOut) {
+        fulfill();
+        return;
+      }
+      const success = predicate.apply(null, args);
+      if (success)
+        fulfill(success);
+      else
+        requestAnimationFrame(onRaf);
+    }
+  }
+
+  /**
+   * @param {number} pollInterval
+   * @return {!Promise<*>}
+   */
+  function pollInterval(pollInterval) {
+    let fulfill;
+    const result = new Promise(x => fulfill = x);
+    onTimeout();
+    return result;
+
+    function onTimeout() {
+      if (timedOut) {
+        fulfill();
+        return;
+      }
+      const success = predicate.apply(null, args);
+      if (success)
+        fulfill(success);
+      else
+        setTimeout(onTimeout, pollInterval);
+    }
+  }
+}";
+
+        internal WaitTask(Frame frame, string predicateBody, string polling, int timeout, object[] args)
+        {
+            _frame = frame;
+            _predicateBody = $"return ( {predicateBody} )(...args)"; //predicateBody;
```

Comment:
not sure we need a conditional operator here since `predicateBody` is always a string

Reply:
We can't validate that the predicate is valid, but maybe we can throw an `InvalidArgumentException` if null or empty.